Exynos Backend for Executorch to bring up on Exynos SoC

Jiseong-oh opened the pull request:
### Summary
This PR introduces support for the Exynos backend of Executorch, can be inferenced on Exynos acceletors.
Exynos Backend is a crucial component designed to optimize the performance of AI models on Exynos chipsets. Leveraging Exynos AI LiteCore, it enables comprehensive support for NPU/DSP, CPU, and GPU, ensuring seamless execution of AI workloads on Exynos-based devices.

Below is a description of the features:
* Exynos Backend : The backends/samsung directory were made by ENN Partitioner, I/F for communication with Exynos LiteCore, runtime backend files and device-specific optimization. See backend/samsung/README.md for usage

* Exynos examples: The examples/samsung directory privides scripts for AOT compilation, quantization and so on. please refer examples/samsung/README.md


### Test plan
This PR is tested with Exynos Backend on NPU/DSP/CPU/GPU of Exynos 2500 SoC. To run the layer and models tests with README.md


cc: @mergennachin @SS-JIA  @digantdesai @kimishpatel

Jiseong-oh committed:
Initial skeleton codes for Exynos Backend

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Provide a python interface for compliation

- Provides a Python interface for compiling NN models via LiteCore in ahead-of-time compilation.

Co-authored-by: chong-chen <chong.chen@samsung.com>
Signed-off-by: chong-chen <chong.chen@samsung.com>
Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Add wrappers for samsung backend

Add Op wrapper and tensor wrapper for Converter of LiteCore.
Also, include the op param wrapper and quantize param wrapper.

Co-authored-by: chong-chen <chong.chen@samsung.com>
Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Serialization from EXIR

- For Exynos AI LiteCore, EXIR will be serialized and converted.

Signed-off-by: chong-chen <chong.chen@samsung.com>
Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Support to converter Exynos's own IR from EXIR

- It can be converted op or graph

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Implement Exynos backend

Implement backend and executor for exynos.

Co-authored-by: chong-chen <chong.chen@samsung.com>
Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
update the license

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Make enn executor for enn backend

- Add the enn executor and backends

Signed-off-by: chong-chen <chong.chen@samsung.com>
Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
add README.md file

- it is descripted how to compile/use Exynos backend.

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Add CMakefile for build

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
add build script for exynos backend

- This script can be supported for building exynos backend easily

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh commented:
@pytorchbot label "release notes: Exynos" "partner:Samsung"

Jiseong-oh committed:
fix: fix backend Name

- update class name

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Add exynos backend for executorch's make/README.md

- Add exynos backend on executorch

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

JacobSzwejbka commented on examples/samsung/executor_runner/enn_executor_runner.cpp:
```diff
@@ -0,0 +1,274 @@
+  fout.close();
+}
+
+int main(int argc, char** argv) {
```
If you dont feel like you need to be able to have your runner execute on baremetal you could use extension/Module rather then dealing with the lower level ET apis like dataloader, method, methodmeta etc

JacobSzwejbka replied:
Ah Im not sure we do have any runner's written against these apis. xnnpack / portable's runner predates there existence and QNN's seems copy pasted from that one with some QNN specific changes. 

In general they should be pretty easy to use though I think. Just need this constructor https://github.com/pytorch/executorch/blob/main/extension/module/module.h#L63 
and then call 
https://github.com/pytorch/executorch/blob/main/extension/module/module.h#L379

extension/tensor is just to dramatically simplify creating ET tensors. Its an owning wrapper that manages the buffers that back the metadata like sizes and strides. [from_blob](https://github.com/pytorch/executorch/blob/main/extension/tensor/tensor_ptr_maker.h#L182) is probably the api you would want to call from there.

hoon98-choi replied:
@JacobSzwejbka 
We'll try to apply your comment for the next PR. thanks!

JacobSzwejbka commented on backends/samsung/runtime/enn_backend.cpp:
```diff
@@ -0,0 +1,95 @@
+  Error execute(
+      BackendExecutionContext& context,
+      DelegateHandle* handle,
+      EValue** args) const override {
```
Ah this has been changed from EValue** to Span<EValue*> on main

GregoryComer commented on examples/samsung/executor_runner/enn_executor_runner.cpp:
```diff
@@ -0,0 +1,274 @@
+/*
```
If you don't need any special logic, you can update the main CMakeLists to link this backend into the main executor runner, when built. That will save you from needing to maintain a backend-specific runner. Adding the backend target into executorch_backends (as mentioned in another comment) will make this work "out of box".

kimishpatel commented on backends/samsung/README.md:
```diff
@@ -0,0 +1,71 @@
+# ExecuTorch Samsung Exynos Delegate
+
+The subtree contains Exynos delegation implementation for ExecuTorch. The target of delegation
```
```suggestion
The subtree contains Exynos delegate implementation for ExecuTorch. The target of delegation
```

kimishpatel commented on backends/samsung/README.md:
```diff
@@ -0,0 +1,71 @@
+# ExecuTorch Samsung Exynos Delegate
+
+The subtree contains Exynos delegation implementation for ExecuTorch. The target of delegation
+is deploying torch model run with exynos NPU/DSP.
```
```suggestion
is deploying torch model to run with exynos NPU/DSP.
```

Jiseong-oh replied:
Thanks, it's fixed.

kimishpatel commented on backends/samsung/README.md:
```diff
@@ -0,0 +1,71 @@
+is deploying torch model run with exynos NPU/DSP.
+
+This backend is implemented on the top of [EXYNOS_LITECORE](https://soc-developer.semiconductor.samsung.com/global/development/light-core)
+Please prepare the SDK before you start, it is important to code compilation and runtime.
```
How do we do that? Is there a documentation? Or clear readme?

Ideal path is for us to have exynos_litecore as pip package so that you can just do `pip install exynos_litecore`

Jiseong-oh replied:
we'll consider installing the LiteCore library using pip later.

mergennachin replied:
Created an issue to track: https://github.com/pytorch/executorch/issues/14004

kimishpatel commented on backends/samsung/README.md:
```diff
@@ -0,0 +1,71 @@
+- Exynos 2500 (E9955)
+
+### Supported Inference Type
+- Quantized (i8/u8/i16/u16)
```
what about int4? is it under dev?

mergennachin replied:
Created an issue to track: https://github.com/pytorch/executorch/issues/14003

kimishpatel commented on backends/samsung/README.md:
````diff
@@ -0,0 +1,71 @@
+```
+
+### Build AOT Targets
+Generated python artifacts allow user call `Compile` interface to lower a model to ENN backend in python script.
````
```suggestion
Generates python artifacts that allow user call `compile` interface to lower a model to ENN backend in python script.
```

Jiseong-oh replied:
Thanks, it's fixed.

GregoryComer commented on backends/samsung/CMakeLists.txt:
```diff
@@ -0,0 +1,141 @@
+            portable_ops_lib
+    )
+    set_target_properties(enn_executor_runner PROPERTIES CXX_VISIBILITY_PRESET hidden)
+    install(
```
To make this work with CMake's find_package functionality, you can add an EXPORT directive here. See https://github.com/pytorch/executorch/blob/d760b772034bfedf49d9fefed2911e63449af623/backends/xnnpack/CMakeLists.txt#L124, for an example. This will also be necessary to add this backend target to the executorch_backends target.

kimishpatel commented on backends/samsung/README.md:
````diff
@@ -0,0 +1,71 @@
+```bash
+./backends/samsung/build.sh -b android --ndk ${ANDROID_NDK}
+```
+ANDROID_ABI=arm64-v8a is default, necessary runtime executable generate in `build_exynos_android` directory.
````
```suggestion
ANDROID_ABI=arm64-v8a is default, necessary runtime executable generated in `build_exynos_android` directory.
```

kimishpatel replied:
what does it generate

mergennachin commented on backends/samsung/README.md:
````diff
@@ -0,0 +1,71 @@
+### Build AOT Targets
+Generated python artifacts allow user call `Compile` interface to lower a model to ENN backend in python script.
+```bash
+./backends/samsung/build.sh -b x86_64
````
Do you wanna add a basic end-to-end integration test to our CI? Can we run everything in x86 environment by any chance?

See example test cases in https://github.com/pytorch/executorch/blob/main/.github/workflows/pull.yml (runs on every PR before merging with main) and https://github.com/pytorch/executorch/blob/main/.github/workflows/trunk.yml (jobs that runs after the PR merges with main)

Jiseong-oh replied:
> > i think you also need to get samsung device for testing CI.
> 
> yeah, we can look into it. also, is there an emulator on linux (either x86 or aarch64) or mac for exynos?

Unfortunately, we don't have the emulator. Could we discuss on slack for this?

Jiseong-oh replied:
> I still think we should add basic unit test, at least for ahead-of-time stuff such as partitioner and serialization. i suppose they won't require samsung phones, right?
> 
> for e2e, we can do it after the PR lands as a fast-follow-on.
you mean we are going to test only until the PTE is generated, right?

mergennachin replied:
For e2e integration test, created an issue to track: https://github.com/pytorch/executorch/issues/14002

GregoryComer commented on backends/samsung/utils/export_utils.py:
```diff
@@ -0,0 +1,29 @@
+)
+
+
+def to_edge_transform_and_lower_to_enn(
```
Do you need this function? I know a few other backends have a similar wrapper, but it would be ideal to use the standard to_edge_transform_and_lower, if possible. If you have any functionality that requires a custom wrapper, we can definitely try to figure out how to extend the core APIs to facilitate that.

digantdesai commented:
I assume we don't have any op support yet? Is that accurate?

kimishpatel commented on backends/samsung/serialization/compile_options_def.fbs:
```diff
@@ -0,0 +1,30 @@
+//============================================================================
```
how is enn artifact serialized? wondering why you need htis one separately

kimishpatel commented on examples/samsung/aot_compiler.py:
```diff
@@ -0,0 +1,84 @@
+    gen_samsung_backend_compile_spec,
+)
+from executorch.backends.samsung.utils.export_utils import (
+    to_edge_transform_and_lower_to_enn,
```
this function `to_edge_transform_and_lower_to_enn` seems like a convenience function, right? And this file aot_compiler.py is really a binary so probably want to call it aot_compile.py

kimishpatel commented on examples/samsung/executor_runner/enn_executor_runner.cpp:
```diff
@@ -0,0 +1,274 @@
+    ET_CHECK_MSG(
+        input_data_reader.nbytes(input_index) == tensor_meta->nbytes(),
+        "Given inputs size is invalid");
+    TensorImpl impl = TensorImpl(
```
@JacobSzwejbka can you point them to a more canonical way of doing this either from llm runner, or elsewhere?

JacobSzwejbka replied:
https://github.com/pytorch/executorch/blob/main/extension/tensor/test/tensor_ptr_maker_test.cpp#L90 

sure an example call to from_blob is here. Its a little awkward I guess because they are getting the shapes from the Meta instead of from the input file in this runner.

kimishpatel commented on examples/samsung/README.md:
```diff
@@ -0,0 +1,59 @@
+# Exynos backend Examples
```
It would be good to provide an example with an actual model as well, in this readme

Jiseong-oh replied:
I added one line about the example model. As you know, we are already supporting many models. We are making sperated commits from the our developed codes to be reviewed easily. In this PR, we update codes only a few models to work.

chong-chen01 committed:
Complete the runtime pipeline

1. Add enn_type.h header file for compilation
2. Complete the pipeline of samsung runtime
3. Remove debug_option in compile options

Co-authored-by: chong-chen <chong.chen@samsung.com>
Co-authored-by: Hoon-Choi  <hoon98.choi@samsung.com>

chong-chen01 committed:
Make partitioner and preprocess work

1. Make compilation pass (script build.sh)
2. Enable simple function for partitioner
3. Fix errors in preprocess phase.

Co-authored-by: chong-chen <chong.chen@samsung.com>

chong-chen01 committed:
Initialize common ops and support ic3/resnet models

Initialize the ops in inception v3 and resnet.
Keep necessary codes in aot_compiler. Use the aot_compiler to generate
ic3, like
```
  python -m executorch.examples.samsung.aot_compiler --chipset e9955 -m ic3
```
About the on-device test, more information can be found in README which
is in the same level directory of aot_compiler.py

Co-authored-by: chong-chen <chong.chen@samsung.com>

Jiseong-oh committed:
change class Name Enn backend

- EnnBackend is an abbreviation for ExynosBackend

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
update README.md file

- updated README.md file according to review by Executorch team

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
add exynos backend to _executorch_backends

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
fix argument data type of main

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
apply updating makefiles

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh commented on CMakeLists.txt:
```diff
@@ -576,6 +576,10 @@ if(EXECUTORCH_BUILD_QNN)
 endif()
 
+if(EXECUTORCH_BUILD_SAMSUNG)
+  add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/backends/samsung)
```
Thanks. I add enn_backend to _executorch_backends.

Jiseong-oh committed:
fixed lint issues

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
update copyright for node_visitor

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh commented on backends/samsung/README.md:
```diff
@@ -0,0 +1,72 @@
+# ExecuTorch Samsung Exynos Delegate
```
Could you more detail explain about this with example? you mean it means that each API used in the Exynos backend needs to be marked as experimental.

mergennachin replied:
> Should we add basic unit test to this Initial PR? If it needs to be added, could you please provide a guide on how to add CI to executorch via email or slack?

Yes, adding basic c++ unit test into executorch would be great.

You can see how other jobs are implemented in https://github.com/pytorch/executorch/blob/main/.github/workflows/pull.yml

You can create a new job, let's say, `unittest-samsung-exynos` and call scripts. There are a few similar examples there.

Jiseong-oh replied:
> > you mean it means that each API used in the Exynos backend needs to be marked as experimental.
> 
> If it's a user facing api (either python or C++), mark it as experimental.

Got it. we will mark it if some api is facing to user.

mergennachin replied:
> It just add codes for CI now. 

Thanks @Jiseong-oh 

I'm seeing this error in the latest CI run

```
++ awk -F ' ' '{print $2}'
++ tr -d '\n'
+ TORCH_VERSION=2.9.0a0+gite7152ff
+ [[ 2.9.0a0+gite7152ff != *\+\g\i\t* ]]
+ source .ci/scripts/setup-samsung-linux-deps.sh
++ set -ex
++ AI_LITE_CORE_VERSION=0.5.0
++ download_ai_lite_core 0.5.0
++ AI_LITE_CORE_VERSION=0.5.0
++ _exynos_ai_lite_core_url='https://ko.ai-studio-farm.com:26310/api/v1/buckets/lite-core/objects/download?prefix=05/ubuntu2204/exynos-ai-litecore-v0.5.0,tar.gz'
++ _exynos_lite_core_dir=/tmp/exynos_ai_lite_core
++ mkdir -p /tmp/exynos_ai_lite_core
++ _tmp_archive=/tmp/exynos-ai-litecore-v0.5.0.tar.gz
++ curl --silent --show-error --location --fail --retry 3 --output /tmp/exynos-ai-litecore-v0.5.0.tar.gz 'https://ko.ai-studio-farm.com:26310/api/v1/buckets/lite-core/objects/download?prefix=05/ubuntu2204/exynos-ai-litecore-v0.5.0,tar.gz'
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
Error: Process completed with exit code 1.
```

it could be that our ubuntu docker image certificates might be old. 

Could you try adding 

```
sudo update-ca-certificates
```

before your curl command?

mergennachin replied:
Also I see that you added python unit tests

Looks like you have to add this line somewhere in test-samsung-models-linux

(It won't be run as part of default unittest job since it is not in pytest.ini)

```
python -m unittest discover backends/samsung/test/ops/ -p "test_*.py"
```

Jiseong-oh replied:
okay, I add the command before curl command.

mergennachin replied:
Lastly, in your CI job, perhaps validate that this command run successfully.

```
        for model in $models; do
          python -m executorch.examples.samsung.aot_compiler --model_name=$model -c E9955
        done
```

mergennachin replied:
cc @chong-chen01 @Jiseong-oh

Jiseong-oh replied:
Yes, we will check if that command for test is running successfully. It should be worked well if downloading of LiteCore is works.

chong-chen01 committed:
Add SamsungTester and build basic test pipeline

Add SamsungTester which provide basic function of lowering module.
Add test for some common ops. Currently, export, partition, lowering to enn backend,
to executorch are included, on-device test and evaluation is skipped.

Outputs of command : python -m unittest discover backends/samsung/test/ops/ -p "test_*.py"
```
INFO:root:Find 1 subgraphs to partition and lowering.
INFO:root:Visiting: aten_add_tensor, aten.add.Tensor
WARNING: Logging before InitGoogleLogging() is written to STDERR
.INFO:root:Find 1 subgraphs to partition and lowering.
INFO:root:Visiting: aten_mul_tensor, aten.mul.Tensor
.
----------------------------------------------------------------------
Ran 17 tests in 15.939s

OK
```

Co-authored-by: chong-chen <chong.chen@samsung.com>

chong-chen01 committed:
Add test cases for samsung enn backend in ci

Currently, support to delegate to EnnBackend for several models.
Enable basic float model like ic3, resnet, etc.

Co-authored-by: chong-chen <chong.chen@samsung.com>

Jiseong-oh committed:
fix ci issue

- fixed certification issue

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh commented on .ci/scripts/setup-samsung-linux-deps.sh:
```diff
@@ -0,0 +1,44 @@
+
+  sudo update-ca-certificates
+
+  curl --silent --show-error --location --fail --retry 3 \
+    --output "${_tmp_archive}" "${_exynos_ai_lite_core_url}"
```
The link to that download is from a temporary server, and the link to the server that can be used as a formula will likely be available starting Friday. Can we temporarily use the --insecure option to first check the behavior for the CI and then apply the code for the official link to the next PR?

mergennachin replied:
@Jiseong-oh Sounds good

Jiseong-oh committed:
add --insecure option

- this option will be revmoed after stable link is defined.

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Merge branch 'main' into exynos-backend

Jiseong-oh committed:
fix LD_LIBRARY_PATH bound error

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Merge branch 'exynos-backend' of https://github.com/Jiseong-oh/executorch into exynos-backend

Jiseong-oh committed:
fix issue library path is reset due to set u

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
ci: change docker image for ci

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

mergennachin commented on .github/workflows/pull.yml:
```diff
@@ -866,6 +866,38 @@ jobs:
+      id-token: write
+      contents: read
+    with:
+      runner: linux.2xlarge
```
You need to change this to linux.arm64.2xlarge

linux.2xlarge is x86 is based machine. The docker image and machine don't match with your configuration.

Jiseong-oh replied:
I changed to other docker image

Jiseong-oh committed:
fixed ci machine of runner

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Merge branch 'pytorch:main' into exynos-backend

Jiseong-oh committed:
Merge branch 'exynos-backend' of https://github.com/Jiseong-oh/executorch into exynos-backend

Jiseong-oh committed:
fix ci issues to setup environment

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Fixed CI issue which Download NDK before compile Exynos backend

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
fix ci issue due to permission

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
fix wrong param issue

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
add testcase for ops

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh commented on .github/workflows/pull.yml:
```diff
@@ -866,6 +866,38 @@ jobs:
+
+        # Setup Samsung SDK (AI Lite Core) and install enn backend
+        source .ci/scripts/setup-samsung-linux-deps.sh
+
```
Hi @mergennachin , I add the testcase in CI.

mergennachin commented on .ci/scripts/setup-samsung-linux-deps.sh:
```diff
@@ -0,0 +1,57 @@
+}
+
+install_enn_backend() {
+  NDK_INSTALLATION_DIR=/opt/ndk
+  rm -rf "${NDK_INSTALLATION_DIR}" && sudo mkdir -p "${NDK_INSTALLATION_DIR}"
+  ANDROID_NDK_VERSION=r27b
+
+  pushd .
+  cd /tmp
+  curl -Os --retry 3 "https://ossci-android.s3.amazonaws.com/android-ndk-${ANDROID_NDK_VERSION}-linux.zip"
+  unzip -qo "android-ndk-${ANDROID_NDK_VERSION}-linux.zip"
+
+  # Print the content for manual verification
+  ls -lah "android-ndk-${ANDROID_NDK_VERSION}"
+  sudo mv "android-ndk-${ANDROID_NDK_VERSION}"/* "${NDK_INSTALLATION_DIR}"
+  popd
+  # build Exynos backend
+  export ANDROID_NDK_ROOT=${ANDROID_NDK_ROOT:-/opt/ndk}
```
We have another docker image that already contains android ndk so that you don't have do download and install. 

You'll have faster CI turnaround time. Here's how docker image is configured


https://github.com/pytorch/executorch/blob/cec1400869e6c1257eba2be2f2b855e0044b6507/.github/workflows/docker-builds.yml#L43

https://github.com/pytorch/executorch/blob/cec1400869e6c1257eba2be2f2b855e0044b6507/.ci/docker/build.sh#L59-L64

https://github.com/pytorch/executorch/blob/cec1400869e6c1257eba2be2f2b855e0044b6507/.ci/docker/ubuntu/Dockerfile#L79-L83

https://github.com/pytorch/executorch/blob/cec1400869e6c1257eba2be2f2b855e0044b6507/.ci/docker/common/install_android.sh#L4

mergennachin commented on backends/samsung/README.md:
````diff
@@ -0,0 +1,71 @@
+
+### Build Anroid Extension
+This is later exposed Java app. Please turn on CMake option `EXECUTORCH_BUILD_ENN`, and ENN runtime will be added.
+```bash
+cmake extension/android \
+  -DCMAKE_TOOLCHAIN_FILE="${ANDROID_NDK}/build/cmake/android.toolchain.cmake" \
+  -DANDROID_ABI="${ANDROID_ABI}" \
+  -DCMAKE_INSTALL_PREFIX=cmake-android-out \
+  -Bcmake-android-out/extension/android
````
cc @Jiseong-oh please take a look at Kimish's question

Jiseong-oh replied:
Fixed it.

mergennachin reviewed (commented):
I created a project board to track all issues related to

https://github.com/orgs/pytorch/projects/140/views/3

Each issue contains a milestone (e.g., 1.0.0, 1.1.0)

Majority of them can be done for 1.1.0 after this PR lands. The only one that I want to land before 9/12 branch cut is https://github.com/pytorch/executorch/issues/13999

Jiseong-oh committed:
ci: remove insecure option for downloading LiteCore

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
fixed target_path for ci

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Merge branch 'main' into exynos-backend

Jiseong-oh committed:
fixed CMake option name

Signed-off-by: jiseong.oh <jiseong.oh@samsung.com>

Jiseong-oh committed:
Merge branch 'exynos-backend' of https://github.com/Jiseong-oh/executorch into exynos-backend

Jiseong-oh commented:
> I created a project board to track all issues related to
> 
> https://github.com/orgs/pytorch/projects/140/views/3
> 
> Each issue contains a milestone (e.g., 1.0.0, 1.1.0)
> 
> Majority of them can be done for 1.1.0 after this PR lands. The only one that I want to land before 9/12 branch cut is #13999

Could't I have a chance to merge next pr before 9/12 branch cut? Acutally, I am preparing to next pr for supporting other float models and ops now.

mergennachin reviewed (approved):
@Jiseong-oh  -- 
1. Thanks for the CI fix, we will try to land this PR now. 
2. Also, you have full access to the project board (https://github.com/orgs/pytorch/projects/140/views/3). You can create new issues and add to the board as you fit. I'd encourage you to proactively use the board to communicate issues and plans, such as what needs to be done as follow-up PR before branch cut, what can be done afterwards etc.

@SS-JIA:

1. please try to reproduce. you probably don't have exynos device, but at least reproduce the AOT part.
4. please import internally and double check it doesn't break any internal test case. Probably it won't since it is an isolated change, but double checking would be good.

cc @kimishpatel , @digantdesai , @JacobSzwejbka @GregoryComer -- any thoughts? if you have major feedback perhaps you can just create an issue in this board (https://github.com/orgs/pytorch/projects/140/views/3), with appropriate milestone and priority attached and they can do as follow-up PRs, if necessary.

SS-JIA commented:
I checked out the PR and was able to build all the components. I was able to build the ENN executor runner and at least run it with `--help` via adb shell. `./install_executorch.sh` works as expected which is good.

However, I was not able to reproduce exporting models with the `aot_compiler.py` script. The first time I ran it, I got

```
/home/ssjia/Github/executorch/src/executorch/backends/samsung/serialization/compile_options.py:14: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  import pkg_resources
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/ssjia/Github/executorch/examples/samsung/aot_compiler.py", line 17, in <module>
    from executorch.backends.samsung.utils.export_utils import (
  File "/home/ssjia/Github/executorch/src/executorch/backends/samsung/utils/export_utils.py", line 11, in <module>
    from executorch.backends.samsung.partition.enn_partitioner import EnnPartitioner
  File "/home/ssjia/Github/executorch/src/executorch/backends/samsung/partition/enn_partitioner.py", line 9, in <module>
    import executorch.backends.samsung.builders.node_visitor as node_visitor
  File "/home/ssjia/Github/executorch/src/executorch/backends/samsung/builders/__init__.py", line 7, in <module>
    from . import (
  File "/home/ssjia/Github/executorch/src/executorch/backends/samsung/builders/node_visitor.py", line 16, in <module>
    from executorch.backends.samsung.serialization.enn_graph_schema import EnnGraph
  File "/home/ssjia/Github/executorch/src/executorch/backends/samsung/serialization/enn_graph_schema.py", line 10, in <module>
    import executorch.backends.samsung.python.PyGraphWrapperAdaptor as PyGraphWrapper
ImportError: libglog.so.0: cannot open shared object file: No such file or directory
```

It seems that there is a dependency on glog/gflags. To attempt to workaround it, I tested building glog/gflags from source (I initially tried to install via `dnf` but no luck). Unfortunately, I now ran into a segmentation fault:

```
[R5CY21Y3WCJ]ssjia@devvm20123 ~/Github/executorch (samsung_test)$ python -m examples.samsung.aot_compiler --chipset e9955 -m ic3
/home/ssjia/Github/executorch/src/executorch/backends/samsung/serialization/compile_options.py:14: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  import pkg_resources
[1]    1423484 segmentation fault (core dumped)  https_proxy=http://fwdproxy:8080 http_proxy=http://fwdproxy:8080 no_proxy=  -
```

Using `lldb`, I found that the seg fault is likely coming from `glog`:

```
>>> runpy.run_module("examples/samsung/aot_compiler")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "<frozen runpy>", line 222, in run_module
  File "<frozen runpy>", line 142, in _get_module_details
ImportError: No module named examples/samsung/aot_compiler
>>> runpy.run_path("examples/samsung/aot_compiler.py")
/home/ssjia/Github/executorch/src/executorch/backends/samsung/serialization/compile_options.py:14: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  import pkg_resources
LLDB diagnostics will be written to /tmp/ssjia/diagnostics-23c32c
Please include the directory content when filing a bug report
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: lldb
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  lldb                 0x00005624f6afb9bb
1  lldb                 0x00005624f6af994b
2  lldb                 0x00005624f6afca3a
3  libc.so.6            0x00007fed9e044560
4  libc.so.6            0x00007fed9e09c993 pthread_kill + 67
5  libc.so.6            0x00007fed9e0444ad raise + 29
6  liblldb.so.15        0x00007feda11c6462
7  libc.so.6            0x00007fed9e044560
8  libglog.so.0         0x00007fed87fdc809
9  libglog.so.0         0x00007fed87fdd346
10 ld.so                0x00007feda4c548af
11 ld.so                0x00007feda4c549a9
12 libc.so.6            0x00007fed9e179a25 _dl_catch_exception + 229
13 ld.so                0x00007feda4c59387
14 libc.so.6            0x00007fed9e1799d0 _dl_catch_exception + 144
15 ld.so                0x00007feda4c58929
16 libc.so.6            0x00007fed9e0968e0
17 libc.so.6            0x00007fed9e1799d0 _dl_catch_exception + 144
18 libc.so.6            0x00007fed9e179a95 _dl_catch_error + 37
19 libc.so.6            0x00007fed9e096433
20 libc.so.6            0x00007fed9e096968 dlopen + 56
21 libpython3.12.so.1.0 0x00007fed9dceeb64
22 libpython3.12.so.1.0 0x00007fed9dcee397
23 libpython3.12.so.1.0 0x00007fed9dcedfb4
24 libpython3.12.so.1.0 0x00007fed9dd59246
25 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
26 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
27 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
28 libpython3.12.so.1.0 0x00007fed9dac0ea2 _PyEval_EvalFrameDefault + 41458
29 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
30 libpython3.12.so.1.0 0x00007fed9dd22e5d
31 libpython3.12.so.1.0 0x00007fed9ddcec74
32 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
33 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
34 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
35 libpython3.12.so.1.0 0x00007fed9dac0ea2 _PyEval_EvalFrameDefault + 41458
36 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
37 libpython3.12.so.1.0 0x00007fed9dd22e5d
38 libpython3.12.so.1.0 0x00007fed9ddcec74
39 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
40 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
41 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
42 libpython3.12.so.1.0 0x00007fed9db34720
43 libpython3.12.so.1.0 0x00007fed9ddcec74
44 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
45 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
46 libpython3.12.so.1.0 0x00007fed9db34ea6 PyImport_ImportModuleLevelObject + 1254
47 libpython3.12.so.1.0 0x00007fed9dac0ea2 _PyEval_EvalFrameDefault + 41458
48 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
49 libpython3.12.so.1.0 0x00007fed9dd22e5d
50 libpython3.12.so.1.0 0x00007fed9ddcec74
51 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
52 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
53 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
54 libpython3.12.so.1.0 0x00007fed9db34720
55 libpython3.12.so.1.0 0x00007fed9ddcec74
56 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
57 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
58 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
59 libpython3.12.so.1.0 0x00007fed9dac0ea2 _PyEval_EvalFrameDefault + 41458
60 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
61 libpython3.12.so.1.0 0x00007fed9dd22e5d
62 libpython3.12.so.1.0 0x00007fed9ddcec74
63 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
64 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
65 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
66 libpython3.12.so.1.0 0x00007fed9dac0ea2 _PyEval_EvalFrameDefault + 41458
67 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
68 libpython3.12.so.1.0 0x00007fed9dd22e5d
69 libpython3.12.so.1.0 0x00007fed9ddcec74
70 libpython3.12.so.1.0 0x00007fed9dac5ba1 _PyEval_EvalFrameDefault + 61169
71 libpython3.12.so.1.0 0x00007fed9db1359c PyObject_CallMethodObjArgs + 556
72 libpython3.12.so.1.0 0x00007fed9db353f2 PyImport_ImportModuleLevelObject + 2610
73 libpython3.12.so.1.0 0x00007fed9dac0ea2 _PyEval_EvalFrameDefault + 41458
74 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
75 libpython3.12.so.1.0 0x00007fed9dd22e5d
76 libpython3.12.so.1.0 0x00007fed9ddcec74
77 libpython3.12.so.1.0 0x00007fed9da52349 PyObject_Vectorcall + 89
78 libpython3.12.so.1.0 0x00007fed9dac3af8 _PyEval_EvalFrameDefault + 52808
79 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
80 libpython3.12.so.1.0 0x00007fed9dd22e5d
81 libpython3.12.so.1.0 0x00007fed9dac4dfe _PyEval_EvalFrameDefault + 57678
82 libpython3.12.so.1.0 0x00007fed9dc54518 PyEval_EvalCode + 152
83 libpython3.12.so.1.0 0x00007fed9dc54353
84 libpython3.12.so.1.0 0x00007fed9dd874e6 PyRun_SimpleStringFlags + 246
85 liblldb.so.15        0x00007feda1024a5e
86 liblldb.so.15        0x00007feda0a900a3
87 liblldb.so.15        0x00007feda0b765bd
88 liblldb.so.15        0x00007feda08a262d lldb::SBDebugger::RunCommandInterpreter(bool, bool) + 157
89 lldb                 0x00005624f6adcc9f
90 lldb                 0x00005624f6add8b5
91 libc.so.6            0x00007fed9e02c657
92 libc.so.6            0x00007fed9e02c718 __libc_start_main + 136
93 lldb                 0x00005624f6ad9491
[1]    1439721 segmentation fault (core dumped)  lldb
```

I raised this with @Jiseong-oh, and it seems that the `glog` dependency issues should be fixed eventually.

I'll merge this PR for now so that Samsung folks can begin working on the next PR. Getting access to a Ubuntu machine to properly reproduce may not be easy, please let me know if anyone has any ideas.

jdwang6666-art commented on backends/samsung/runtime/enn_api_implementation.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+Error EnnApi::loadApiLib() {
+  const char enn_api_lib_name[] = "libenn_public_api_cpp.so";
```
Where can i get libenn_public_api_cpp.so? MY LOG is HERA:
dlopen failed: dlopen failed: library "/data/local/tmp/wjd/SamsungTest/libenn_public_api_cpp.so" not found
